src/log.rs

```diff
@@ -0,0 +1,154 @@
+///
+/// * If the first bit is not set AND the second bit is set, this is
+///   an event with payload. Treat the next item in the stream as that
+///   payload.
+#[derive(Clone, Copy, PartialEq)]
+#[repr(transparent)]
+pub struct LogEntry(u32);
+
+impl LogEntry {
+    /// Construct an entry without knowing what kind it is.
+    ///
+    /// # Safety
+    ///
+    /// This should be used with caution, for cases where you
+    /// want to use `LogEntry` to determine the "type" of the raw
+    /// value.
+    pub unsafe fn new_unchecked(val: u32) -> Self {
+        LogEntry(val)
+    }
+
+    /// Create a single event `LogEntry` with no payload.
+    #[must_use]
+    #[inline]
+    pub fn event(event_id: EventId) -> Self {
+        // The construction checks for EventId should prevent the top bit from being set
+        LogEntry(event_id.get_raw())
+    }
+
+    /// Create a logical-clock style pair of `LogEntry`s
+    #[must_use]
+    #[inline]
+    pub fn clock(clock: LogicalClock) -> (Self, Self) {
+        // Set the top bit for id to indicate that it is not an event but a logical clock bucket,
+        // and to treat the next item as the count. Don't separate these two!
+        let id = clock.id.get_raw() | CLOCK_MASK;
+        (
+            LogEntry(id),
+            LogEntry(pack_clock_word(clock.epoch, clock.ticks)),
+        )
+    }
+
+    /// Create a pair of `LogEntry`s representing an event and
+    /// associated payload.
+    #[must_use]
+    #[inline]
+    pub fn event_with_payload(event_id: EventId, payload: u32) -> (Self, Self) {
+        (
+            LogEntry(event_id.get_raw() | EVENT_WITH_PAYLOAD_MASK),
+            // We're just giving payload back out as is for now.
+            LogEntry(payload),
+        )
+    }
+
+    /// Determine if the clock bit is set on this entry.
+    #[inline]
+    pub fn has_clock_bit_set(self) -> bool {
+        (self.0 & CLOCK_MASK) == CLOCK_MASK
+    }
+
+    /// Determine if the event with payload bit is set on this entry.
+    #[inline]
+    pub fn has_event_with_payload_bit_set(self) -> bool {
+        (self.0 & EVENT_WITH_PAYLOAD_MASK) == EVENT_WITH_PAYLOAD_MASK
+    }
+
+    /// Get the underlying value as a convenient primitive
+    #[inline]
+    pub fn raw(self) -> u32 {
+        self.0
+    }
+
+    /// Unset that top bit to get the original probe id back out
+    #[inline]
+    pub fn interpret_as_logical_clock_probe_id(self) -> u32 {
+        self.0 & !CLOCK_MASK
+    }
+
+    /// Convert to an event id.
+    #[inline]
+    pub fn interpret_as_event_id(&self) -> Option<EventId> {
+        let id = self.0 & !EVENT_WITH_PAYLOAD_MASK;
+        EventId::new(id).or_else(|| EventId::new_internal(id))
+    }
+}
+
+impl core::fmt::Debug for LogEntry {
+    #[inline]
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+        f.write_fmt(format_args!("LogEntry({})", self.0))
+    }
+}
+
+impl Entry for LogEntry {
+    fn is_prefix(&self) -> bool {
```

Comment:
This function might still return a misleading value if it is used on a payload / clock value.  Maybe add some warning comment?